### DISCUSSION TITLE
How to use minio with pocketbase?

### DISCUSSION THREAD

**Discussion by Lex-2008** [pocketbase/pocketbase#90] [Q&A] [answered]
I'm trying to use minio with pocketbase (both are pretty new to me, but both are distributed as single executables :+1: )

Following minio guide: https://min.io/download#/linux
```
wget https://dl.min.io/server/minio/release/linux-amd64/minio
chmod +x minio
mkdir minio-data
./minio server minio-data --console-address ":9001"
```
It tells me that it has API at http://127.0.0.1:9000 and console at http://127.0.0.1:9001, among other things

I then open minio console, create a bucket "test" and configure S3 in pocketbase like this:
Endpoint: 127.0.0.1
Bucket: test
Region: x
access key: minioadmin
secret key: minioadmin

But when I try to attach a file to a record in pocketbase, it says "Failed to create record." (in debug mode: "Failed to upload all files").

I suspect that it's because of #78, but setting endpoint to the full URL by modifying ~/pb_data/data.db (with sqlite3, when pocketbase is off) doesn't help.

**Have anyone succeeded in using pocketbase together with minio?**

FWIW, using `s3cmd` works with these arguments:
```
s3cmd --no-ssl --access_key=minioadmin --secret_key=minioadmin --host=127.0.0.1:9000 --host-bucket=127.0.0.1:9000
```

**Comment by Lex-2008** [comment]
Actually, you don't even need to install minio - they provide a "playground" minio instance at play.min.io:9000, access_key is `Q3AM3UQ867SPQQA43P2F`, secret_key is `zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG`.

**Comment by ganigeorgiev** [comment]
@Lex-2008 I haven't tested with Minio (only DO and Vultr S3 offerings) but I suspect that it is some access control policy misconfiguration or similar. I'll change the endpoint to allow URIs and will update the dev error reporting to include the filesystem errors too.

**Comment by Lex-2008** [comment]
Thanks! Indeed it might be some issue on my side, since @gedw99 seems to have them running together.

@gedw99: please let me know when you have something, I'll be happy to look at it!

No rush, though.

**Comment by Lex-2008** [comment]
Thanks @ganigeorgiev for implementing plain URL and error logging! I moved a bit closer to solution, but it still doesn't work for me :)

My findings, if anyone finds them useful:

It looks like pocketbase unconditionally prepends bucket name to endpoint and tries to upload file to `http://test.127.0.0.1:9000/` (at this point, DNS resolution fails). If I change endpoint to hostname and add it to `/etc/hosts` - then pocketbase talks to minio, but still unsuccessfully.

Comparing minio "trace" logs generated by pocketbase and s3cmd uploads, it looks like s3cmd first executes `s3.GetBucketLocation` and then uses hostname without bucketname (i.e. plain `localhost` - probably as instructed by minio), while pocketbase always prepends bucketname to hostname (i.e. connects to `test.localhost`) - what confuses minio.

So it looks like minio is not-s3-enough for AWS library used by pocketbase :)

The option `S3ForcePathStyle` can be related:
https://github.com/aws/aws-sdk-go/blob/main/aws/config.go#L118
but I don't want to push @ganigeorgiev to work on this too much :) especially since AWS themselves deprecated it.

Also, if there's still an option, probably this can be converted to a discussion, since it's not an issue with pocketbase, but actually more like a usage question.

**Comment by Lex-2008** [comment]
oh, I just have to set `MINIO_DOMAIN=localhost`, and it seems to be working!

I'll test a bit more and make a write-up, maybe later today.

**Comment by ekremparlak** [comment]
I was able to use minio with pocketbase without problem. 
When i looked for logs i found the request use virtual hosted-style requests. Like for a bucket named test you can use s3.example.com/test or test.s3.example.com . Pocketbase was using test.s3.example.com . 
It could be a problem with some providers that doesn't support virtual hosted-style requests.
Related article https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html

**Comment by Lex-2008** [comment, marked as answer]
TL;DR:
* use hostname to connect to minio (`http://{hostname}:9000/`),
* ensure `{bucketname}.{hostname}` resolves from where pocketbase is running and points to minio,
* start minio with `MINIO_DOMAIN={hostname}` env variable

Longer explanation follows.

Prereqs:
1. pocketbase and minio servers running on different or a same machine, machine running minio must be reachable from machine running pocketbase.
1. Pick a hostname which pocketbase will use to connect to minio server. It doesn't need to be globally unique, simple `myminioserver` will do.
2. Pick a name for the bucket. I'll use `test` for testing.
3. You need to know IP address of the minio machine (minio usually prints IP addresses it can find when it starts)

Steps.
1. Configure DNS. Simplest way is to edit `/etc/hosts` file on the machine where pocketbase is running. Add two lines to this file:
    ```
    {minio-ip} {minio-hostname}
    {minio-ip} {bucket-name}.{minio-hostname}
    ```
    For example, with minio and pocketbase running on the same machine, and "test" bucket name:
    ```
    127.0.0.1  myminioserver
    127.0.0.1  test.myminioserver
    ```
2. start minio like this
    ```
    MINIO_DOMAIN={minio-hostname} ./minio server {minio-data-dir}
    ```
    where `{minio-data-dir}` is your minio data directory. For example:
    ```
    MINIO_DOMAIN=myminioserver ./minio server ~/minio_data
    ```
    Note that this looks to be pretty insecure: it's started with root username and password set to default value `minioadmin`, and API endpoint exposed to the whole world who can reach it. Properly securing minio goes beyond this guide, but don't tell I didn't warn you!
3. open `http://{minio-ip}:9000` in your browser (minio prints some of reachable IPs during its startup). With minio running on the same machine as browser, http://127.0.0.1:9000 should work. Login as `minioadmin` user with `minioadmin` password. and create a bucket for pocketbase. If you previously decided to call it "test" - that's how you should call it. You may also want to create a separate user for pocketbase, and maybe create a service account for it, but I'll skip this part for simplicity.
5. Configure pocketbase: open http://localhost:8090/_/ if it's running on your local machine, and navigate to [settings->storage](http://localhost:8090/_/?#/settings/storage). Set:
    * endpoint: `http://{minio-hostname}:9000/` - for example, `http://myminioserver:9000/`
    * bucket: bucket name, for example `test`
    * region: anything, for example `x`. In default configuration minio doesn't check it, but if you set it in minio settings (/settings/configurations/region) - then these values must match
    * access key and secret: both `minioadmin` for now, but remember to secure your minio server in the future!
6. that's all! In pocketbase, create a collection with a field of type "file", and try to create a record. If it succeeded - open minio, open the bucket you created for pocketbase, and click "Browse" button - you should see directory structure with your file in it.

Next steps.
To properly secure pocketbase-minio connection, you must:
1. Start minio with a different root password - use `MINIO_ROOT_USER` and `MINIO_ROOT_PASSWORD` env variables for that.
1. give it [certificates](https://docs.min.io/docs/how-to-secure-access-to-minio-server-with-tls.html) and switch from http to https. I'm not 100% sure, but it might be that anyone sniffing traffic between pocketbase and minio might learn your minio password. Or maybe read and modify transferred files.
2. (optional) create a separate user for pocketbase so a compromised pocketbase couldn't affect rest of your minio installation
3. (optional) encrypt your pocketbase settings - use `--encryptionEnv` argument when starting pocketbase - since otherwise your minio credentials (access key and secret) are stored in database unencrypted.

**Reply by regaipserdar** [reply]
Hi @Lex-2008 , 
This works perfectly, thank you. However, I believe we should add this to the documentation.

**Reply by bartcode** [reply]
I can also confirm this works. For anyone still struggling to get this to work on Kubernetes: you might want to update `kube-system/coredns` (`ConfigMap`) and add a rewrite rule. Assuming a `minio` service runs in a namespace called `minio`, this should work:

```text
.:53 {
    errors
    health
    ready
    rewrite name regex (.*)\.minio\.minio\.svc\.cluster\.local minio.minio.svc.cluster.local
    # ...
}
```
